src/plugins/discover/public/context_awareness/profile_providers/log_document_profile/profile.ts

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { DataTableRecord } from '@kbn/discover-utils';
+import { DocumentProfileProvider, DocumentType } from '../../profiles';
+import { ProfileProviderServices } from '../../profiles/profile_provider_services';
+
+export const createLogDocumentProfileProvider = (
+  services: ProfileProviderServices
+): DocumentProfileProvider => ({
+  profileId: 'log-document-profile',
+  profile: {},
+  resolve: ({ record }) => {
+    const isLogRecord = getIsLogRecord(record, services.logsContextService.isLogsIndexPattern);
+
+    if (!isLogRecord) {
+      return { isMatch: false };
+    }
+
+    return {
+      isMatch: true,
+      context: {
+        type: DocumentType.Log,
+      },
+    };
+  },
+});
+
+const getIsLogRecord = (
+  record: DataTableRecord,
+  isLogsIndexPattern: ProfileProviderServices['logsContextService']['isLogsIndexPattern']
+) => {
+  return (
+    getDataStreamType(record).includes('logs') ||
+    hasFieldsWithPrefix('log.')(record) ||
+    getIndices(record).some(isLogsIndexPattern)
+  );
+};
+
+const getFieldValues =
+  (field: string) =>
+  (record: DataTableRecord): unknown[] => {
+    const value = record.flattened[field];
+    return Array.isArray(value) ? value : [value];
+  };
+
+const getDataStreamType = getFieldValues('data_stream.type');
+const getIndices = getFieldValues('_index');
+
+const hasFieldsWithPrefix = (prefix: string) => (record: DataTableRecord) => {
+  return Object.keys(record.flattened).some((field) => field.startsWith(prefix));
```

Comment:
We sometimes deal with records in Discover containing thousands of fields. We're deferring document context resolution until needed anyway, but a `for...of` loops would be more efficient here vs allocating a new array for each document with `Object.keys`.

Reply:
I think you meant using a `for...in` loop over the properties in `record.flattened` to check whether any of them starts with the passed prefix, `for...of` is only for iterables.

In the case of `for...in`, it is natively faster but requires an additional check on the [object's own properties](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/hasOwnProperty) only, which makes the performance almost the same.

For readability I'd rather leave it like this, [here](https://measurethat.net/Benchmarks/ShowResult/529469) you can also find a performance benchmark for the different approaches which highlights a relatively small difference.

Reply:
Yes, definitely meant `for...in` 🙂 and I was thinking a bit more about allocations when we unfortunately encounter docs with something like 80,000 fields, but this optimization might be premature, so it makes sense to leave it as is and look into it further if it ever becomes an issue.